Meta: Update style guide to discuss preferred orderings of elements (#831)
tc39/ecma402: docs/style-guide.md

## Table of Contents

- [Casing Conventions](#casing-conventions)
	- [Identifiers defined by ECMA-401](#identifiers-defined-by-ecma-401)
		- [Examples](#examples)
		- [Alternative: Kebab case for all string enumerations](#alternative-kebab-case-for-all-string-enumerations)
			- [Pros](#pros)
			- [Cons](#cons)
			- [Decision](#decision)
		- [Alternative: Kebab case for new identifiers only](#alternative-kebab-case-for-new-identifiers-only)
			- [Pros](#pros-1)
			- [Cons](#cons-1)
			- [Decision](#decision-1)
		- [Alternative: Use kebab case but also accept camel case](#alternative-use-kebab-case-but-also-accept-camel-case)
			- [Pros](#pros-2)
			- [Cons](#cons-2)
			- [Decision](#decision-2)
	- [Identifiers defined outside ECMA-402](#identifiers-defined-outside-ecma-402)
		- [Examples](#examples-1)
		- [Alternative: Convert identifiers to camel case](#alternative-convert-identifiers-to-camel-case)
			- [Pros](#pros-3)
			- [Cons](#cons-3)

> 			- [Decision](#decision-3)

Reviewer: Are the whitespace changes necessary?

Author: changed it back to the previous spacing

Reviewer: The previous spacing was 4 spaces and your new version has 2 spaces :see_no_evil:

            - [Pros](#pros-3)
            - [Cons](#cons-3)
            - [Decision](#decision-3)

>     - [Element Ordering](#element-ordering)
> 	      - [General Guidelines](#general-guidelines)
> 	      - [resolvedOptions](#resolvedoptions)

Reviewer 2: The extra indentation here breaks Markdown list output: https://github.com/tc39/ecma402/blob/97ba211cddb8a4a22f67e327af05372b545f2bcd/docs/style-guide.md#table-of-contents:~:text=Element%20Ordering
```suggestion
    - [Element Ordering](#element-ordering)
	    - [General Guidelines](#general-guidelines)
	    - [resolvedOptions](#resolvedoptions)
```

##### Decision

- Defer the decision on the syntax for identifiers to the other specification when possible.

> 

Reviewer 3: @ben-allen Please make sure this fully fixes #578

Reviewer 3: Actually we should follow this up with a normative PR for PluralRules.

## Element Ordering

> This section concerns the order in which containers store elements. 

Reviewer 2: What is the definition of "container" in this context?

### General Guidelines

> :star2: ECMA-402 spec must provide a deterministic order for the contents of all containers. This order should be lexicographic except in cases wherein there is a clearly preferable semantic ordering. :star2:

Reviewer 2: ```suggestion
:star2: *ECMA-402 spec must provide a deterministic order for the contents of all containers. This order should be lexicographic except in cases wherein there is a clearly preferable semantic ordering.* :star2:
```

Reviewer 2: Consistency with existing conventions at [Identifiers defined by ECMA-402](https://github.com/tc39/ecma402/blob/ff6c9c68d629d1cdc3edf45755262e15d17e84b1/docs/style-guide.md#identifiers-defined-by-ecma-402) and [Identifiers defined outside ECMA-402](https://github.com/tc39/ecma402/blob/ff6c9c68d629d1cdc3edf45755262e15d17e84b1/docs/style-guide.md#identifiers-defined-outside-ecma-402).

> For an example of when *not* to use lexicographic order, consider an array holding days of the week. This array should store those days in the order `["monday", "tuesday", "wednesday", ...]` instead of the lexicographic `["friday", "monday", "saturday", "sunday", ...]` order.

Reviewer 3: Choose something that is actually in the spec:

```suggestion
For an example of when *not* to use lexicographic order, consider an array holding days of the week. This array should store those days in the order `["years", "months", "weeks", "days"]` instead of the lexicographic `["days", "months", "weeks", "years"]` order.
```

Reviewer 2: ```suggestion
For an example of when *not* to use lexicographic order, consider an array holding calendar time scale units. This array could use descending magnitude as `["years", "months", "weeks", "days"]` or its reverse ascending magnitude, but should not use the lexicographic `["days", "months", "weeks", "years"]` order.
```

Please also provide at least one positive example.

### `resolvedOptions`

:star2: The `resolvedOptions` of Intl objects should appear in the following order: 

1. `locale`
2. All properties (given in lexicographical order) that can be set by extension keys and that are guaranteed to exist
3. Properties (in lexicographical order) that are not set by extension keys and that are guaranteed to exist 

> 4. All properties (in lexicographical order) that exist conditionally. :star2:

Reviewer 3: @gibson042 for feedback on this ordering

Reviewer 2: It seems insufficiently stable to me, because we are currently contemplating adding support for -u- extension keys that can currently only be set by options, and could definitely promote a conditional property to mandatory in the future.

```suggestion
### `resolvedOptions`

:star2: *Properties in the output from `resolvedOptions()` of an Intl object should start with `locale` and otherwise use lexicographic order. Deviations (such as those motivated by semantic order involving a subset of properties) must be documented.* :star2:

Spec changes that add properties should do so in accordance with this recommendation, rather than automatically placing them at the end—relative enumeration order of properties should remain stable over time, but there is no such expectation regarding adjacency.

#### Examples

`Object.keys(new Intl.Segmenter().resolvedOptions())` returns `[ 'locale', 'granularity' ]`. If support for the [Unicode BCP 47 U Extension "dx" key](https://unicode.org/reports/tr35/#UnicodeDictionaryBreakExclusionIdentifier) were added and exposed as `dictionaryBreakExcludedScripts`, that property would belong before `granularity`.
```

### General Guidelines

:star2: *ECMA-402 spec must provide a deterministic order for the contents of all containers. This order should be lexicographic except in cases wherein there is a clearly preferable semantic ordering. :star2:*

For example, an array holding the time zone identifiers `Asia/Tokyo`, `Asia/Ho_Chi_Minh`, and `Asia/Dubai` should use the order `["Asia/Dubai", "Asia/Ho_Chi_Minh", "Asia/Tokyo"]`
  
For an example of when *not* to use lexicographic order, consider an array holding calendar time scale units. This array could be ordered by descending magnitude as `["years", "months", "weeks", "days"]` or its reverse ascending magnitude as `["days", "weeks", "months", "years"]`, but should not use the lexicographic `["days", "months", "weeks", "years"]` order.

### `resolvedOptions`

:star2: <em>The `resolvedOptions` of Intl objects should appear in the following order: 

1. `locale`
2. All properties (given in lexicographical order) that can be set by extension keys and that are guaranteed to exist
3. Properties (in lexicographical order) that are not set by extension keys and that are guaranteed to exist 

> 4. All properties (in lexicographical order) that exist conditionally.</em> :star2:

Reviewer: Is this whole section purposefully italicized?

Author: Yes, to match the italicization in the other starred blocks. I do admit that italicizing the list might be a situation where the result of insisting on consistency is ugly enough that it's the wrong thing to do.

Reviewer: I think one way to do it would be to rephrase the first sentence into a summarized guideline, star+italicize it and follow that with the full list. An entire list being starred/italicized doesn't really say much and looks weird.

##### Decision

- Defer the decision on the syntax for identifiers to the other specification when possible.

> ## Element Ordering
> 
> This section concerns the order in which containers store elements. 
> 
> ### General Guidelines
> 
> :star2: *ECMA-402 spec must provide a deterministic order for the contents of all containers. This order should be lexicographic except in cases wherein there is a clearly preferable semantic ordering. :star2:*

Reviewer 2: ```suggestion
## Element Ordering

This section concerns the observable order for elements of arrays and properties of objects.

### General Guidelines

:star2: *ECMA-402 must provide a deterministic order for the elements of all arrays and the properties of all objects. This order should be lexicographic except in cases for which there is a clearly preferable semantic ordering. :star2:*
```

> For example, an array holding the time zone identifiers `Asia/Tokyo`, `Asia/Ho_Chi_Minh`, and `Asia/Dubai` should use the order `["Asia/Dubai", "Asia/Ho_Chi_Minh", "Asia/Tokyo"]`

Reviewer 2: ```suggestion
For example, an array holding the time zone identifiers "Asia/Tokyo", "Asia/Ho_Chi_Minh", and "Asia/Dubai" should use the order `["Asia/Dubai", "Asia/Ho_Chi_Minh", "Asia/Tokyo"]`
```

  
For an example of when *not* to use lexicographic order, consider an array holding calendar time scale units. This array could be ordered by descending magnitude as `["years", "months", "weeks", "days"]` or its reverse ascending magnitude as `["days", "weeks", "months", "years"]`, but should not use the lexicographic `["days", "months", "weeks", "years"]` order.

> ### `resolvedOptions`
> 
> :star2: *The `resolvedOptions` of `Intl` objects should appear with `locale` first, followed by properties guaranteed to exist that can be set via extension keys, followed by properties guaranteed to exist that are not set by extension keys, and then finally all properties that exist conditionally. Elements in each of these categories should appear in lexicographical order.* :star2:
> 
> Order of `resolvedOptions`:
> 1. `locale`
> 2. All properties (given in lexicographical order) that can be set by extension keys and that are guaranteed to exist
> 3. Properties (in lexicographical order) that are not set by extension keys and that are guaranteed to exist 
> 4. All properties (in lexicographical order) that exist conditionally.

Reviewer 2: Reiterating https://github.com/tc39/ecma402/pull/831#discussion_r1330814261 :
```suggestion
### `resolvedOptions`
:star2: *Properties in the output from `resolvedOptions()` of an Intl object should start with `locale` and otherwise use lexicographic order. Deviations (such as those motivated by semantic order involving a subset of properties) must be documented.* :star2:
Spec changes that add properties should do so in accordance with this recommendation, rather than automatically placing them at the end—relative enumeration order of properties should remain stable over time, but there is no such expectation regarding adjacency.
#### Examples
`Object.keys(new Intl.Segmenter().resolvedOptions())` returns `[ 'locale', 'granularity' ]`. If support for the [Unicode BCP 47 U Extension "dx" key](https://unicode.org/reports/tr35/#UnicodeDictionaryBreakExclusionIdentifier) were added and exposed as `dictionaryBreakExcludedScripts`, that property would belong before `granularity`.
```

Author: I'm revising this one again and though the reasoning in [this comment](https://github.com/tc39/ecma402/pull/831#discussion_r1330814261) seems sound to me, I wanted to verify that we've got agreement that the "put conditionally existing properties after guaranteed to exist properties" language is what we want to go with. @gibson042 @ryzokuken @sffc

Reviewer 3: No strong opinion between the two orderings. I guess I slightly favor putting the extension keywords adjacent in the list to the locale and in such a way that they could be sliced out and fed to the Locale constructor, but that is an error-prone operation, so Richard's ordering is fine, too.
